enterprise_access/apps/content_assignments/models.py

```diff
@@ -246,7 +400,66 @@ def bulk_update(cls, assignment_records, updated_field_names):
             batch_size=BULK_OPERATION_BATCH_SIZE,
         )
 
-    def get_auto_expiration_date(self):
+    def learner_acknowledged(self):
+        """
+        Returns whether or not the learner has acknowledged the assignment.
+        """
+        if self.state == LearnerContentAssignmentStateChoices.EXPIRED:
+            last_expired_action = self.get_last_successful_expiration_action()
+            if not last_expired_action:
+                logger.warning(
+                    'LearnerContentAssignment with UUID %s is in an expired state, but has no related '
+                    'actions in an expired state.',
+                    self.uuid,
+                )
+                return False
+            return last_expired_action.learner_acknowledged()
+
+        if self.state == LearnerContentAssignmentStateChoices.CANCELLED:
+            last_cancelled_action = self.get_last_successful_cancel_action()
+            if not last_cancelled_action:
+                logger.warning(
+                    'LearnerContentAssignment with UUID %s is in a cancelled state, but has no related '
+                    'actions in a cancelled state.',
+                    self.uuid,
+                )
+                return False
+            return last_cancelled_action.learner_acknowledged()
+
+        # Fallback to None, in case the assignment is in a state that may not be acknowledged.
+        return None
+
+    def _get_enrollment_deadline_date(self, content_metadata):
+        """
+        Helper to get the enrollment end date from a content metadata record.
+        """
+        if content_metadata is not None:
+            normalized_metadata = content_metadata.get('normalized_metadata') or {}
+            enrollment_end_date_str = normalized_metadata.get('enroll_by_date')
+            try:
+                datetime_obj = parse_datetime_string(enrollment_end_date_str)
+                if datetime_obj:
+                    return datetime_obj.replace(tzinfo=UTC)
+            except ValueError:
+                logger.warning(
+                    'Bad datetime format for %s, value: %s',
+                    content_metadata.get('key'),
+                    enrollment_end_date_str,
+                )
+        return None
+
+    def get_subsidy_expiration(self):
```

Comment:
I have the same sort of question here about which file this code and `get_automatic_expiration_date_and_reason()` really belongs in.  This stuff touches a couple of different external domains (content metadata, subsidy expiration), I'd lean toward moving this to `api.py` (I don't know how much of a headache this would cause in the serializers, though).
The goal would be to help keep the models layer "thin", so that we don't long-term end up with a tangled web of dependencies in one of the core areas, and let `api.py` be kind of the "octopus" of business logic that has its hands in lots of places. I'm open to a differing view on that, though.

Reply:
Updated. Ended up with other circular import issues related to other methods after moving `get_automatic_expiration_date_and_reason` to the `enterprise_access.apps.api.utils` file. I decided to go up one more level to `enterprise_access.utils` instead for now, which does get around the issue as well.